### PR TITLE
[SPARK-43205][DOCS][SQL][FOLLOWUP] IDENTIFIER clause docs

### PR DESCRIPTION
**Pull request by srielau**
### What changes were proposed in this pull request?

This is a follow-up of #41007 .

This pull request documents the IDENTIFIER clause

### Why are the changes needed?

If we don't document it, it doesn't exist...

### Does this PR introduce _any_ user-facing change?

Yes

### How was this patch tested?

Acceptance test.
Screenshot 2023-08-15 at 5 31 13 PM
Screenshot 2023-08-15 at 5 31 06 PM
Screenshot 2023-08-15 at 5 30 54 PM

### CODE REVIEW

**Comment by srielau** [comment]
@gengliangwang how do we test doc changes?

File: docs/sql-ref-identifier-clause.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+spark.sql("SELECT * FROM IDENTIFIER(:tabname)", args = Map("tabname" -> "myschema.mytable")
+
+-- Using IDENTIFIER() on a column reference 
+spark.sql("SELECT IDENTIFIER(:colname) FROM mychema.mytable WHERE IDENTIFIER(:colname) IS NOT NULL", args = Map("colname" -> "`c1`")
+
+-- Using IDENTIFIER() on a function reference 
+spark.sql("SELECT IDENTIFIER(:aggFunc)(c1) FROM myschema.mytable", args = Map("aggFunc" -> "sum")
+
+-- Using IDENTIFIER() on SHOW
+spark.sql("SHOW TABLES IN IDENTIFIER(:namepace)", args = Map("namespace" -> "mychema")
```
Comment (srielau) [inline docs/sql-ref-identifier-clause.md:69]:
```suggestion
spark.sql("SHOW TABLES IN IDENTIFIER(:namespace)", args = Map("namespace" -> "mychema")
```

File: docs/sql-ref-identifier-clause.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+- qualified table name
+- namespace name
+- function name
+- qualified column or attribute reference
+
+### Examples
+
+```scala
+-- Using IDENTIFIER() on a namepace
+spark.sql("CREATE SCHEMA IDENTIFIER(:namepace)", args = Map("namespace" -> "mychema")
```
Comment (srielau) [inline docs/sql-ref-identifier-clause.md:51]:
```suggestion
spark.sql("CREATE SCHEMA IDENTIFIER(:namespace)", args = Map("namespace" -> "mychema")
```

**Comment by gengliangwang** [comment]
@srielau Yes, please following https://github.com/apache/spark/blob/master/docs/README.md and preview your doc change in the browser.
(Note that the Scala/Python API Prerequisites are not required for this one.)

**Review by MaxGekk — COMMENTED** [review]
@srielau Could you add the `[DOC]` tag to PR's title, and a screenshot of you the web-page w/ your changes (add it PR description).

File: docs/sql-ref-identifier-clause.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+spark.sql("CREATE SCHEMA IDENTIFIER(:namespace)", args = Map("namespace" -> "mychema")
+
+-- Using IDENTIFIER() on table DDL
+spark.sql("CREATE TABLE IDENTIFIER('myschema.' || :tabname) (c1 INT)", args = Map("tabname" -> "mytable")
+
+-- Using IDENTIFIER() on table DML
+spark.sql("INSERT INTO IDENTIFIER(:tabname) VALUES (1)", args = Map("tabname" -> "myschema.mytable")
+
+-- Using IDENTIFIER() on a table reference 
+spark.sql("SELECT * FROM IDENTIFIER(:tabname)", args = Map("tabname" -> "myschema.mytable")
```
Comment (dongjoon-hyun) [inline docs/sql-ref-identifier-clause.md:60]: Could you include the output here?

File: docs/sql-ref-identifier-clause.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+-- Using IDENTIFIER() on a column reference 
+spark.sql("SELECT IDENTIFIER(:colname) FROM mychema.mytable WHERE IDENTIFIER(:colname) IS NOT NULL", args = Map("colname" -> "`c1`")
+
+-- Using IDENTIFIER() on a function reference 
+spark.sql("SELECT IDENTIFIER(:aggFunc)(c1) FROM myschema.mytable", args = Map("aggFunc" -> "sum")
+
+-- Using IDENTIFIER() on SHOW
+spark.sql("SHOW TABLES IN IDENTIFIER(:namespace)", args = Map("namespace" -> "mychema")
```
Comment (dongjoon-hyun) [inline docs/sql-ref-identifier-clause.md:69]: ditto. Please include output too

File: docs/sql-ref-identifier-clause.md (hunk 3 above)
Comment (srielau) [reply docs/sql-ref-identifier-clause.md]: Done

File: docs/sql-ref-identifier-clause.md (hunk 4 above)
Comment (srielau) [reply docs/sql-ref-identifier-clause.md]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala [hunk 5]
```diff
@@ -305,16 +307,16 @@ trait ColumnResolutionHelper extends Logging {
         } else {
           variable
         }
-      }.map(e => Alias(e, nameParts.last)())
+      }.map(e => if (isTopLevel) Alias(e, nameParts.last)() else e )
     }
 
     e.transformWithPruning(_.containsAnyPattern(UNRESOLVED_ATTRIBUTE, TEMP_RESOLVED_COLUMN)) {
```
Comment (linhongliu-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala:313]: how about skip wrapping `Alias` in the `def resolve`, but check the final value after `e.tranformWithPruning`, if it's not an `Alias`, add one.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIdentifierClause.scala [hunk 6]
```diff
@@ -42,7 +42,12 @@ object ResolveIdentifierClause extends Rule[LogicalPlan] {
 
   private def evalIdentifierExpr(expr: Expression): Seq[String] = {
     expr match {
-      case e if !e.foldable => expr.failAnalysis(
+      case e: Alias if !e.child.foldable => expr.failAnalysis(
+        errorClass = "NOT_A_CONSTANT_STRING.NOT_CONSTANT",
+        messageParameters = Map(
+          "name" -> "IDENTIFIER",
+          "expr" -> expr.sql))
+      case e if !e.isInstanceOf[Alias] && !e.foldable => expr.failAnalysis(
```
[the remark this answers is not in the document]
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIdentifierClause.scala]: Removed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala (hunk 5 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala]: Removed

**Comment by srielau** [comment]
@cloud-fan This is also ready I think.

**Comment by cloud-fan** [comment]
the test failure is unrelated, thanks, merging to master!

**Pull request closed without merging**